sidecar/proxy/monitor/registry.go

```diff
@@ -85,44 +92,70 @@ func (m *registry) Start() error {
 func (m *registry) poll() error {
 	// Get newest catalog from registry
 	latestCatalog, err := m.getLatestCatalog()
+
 	if err != nil {
 		logrus.WithError(err).Warn("Could not get latest catalog from registry")
 		return err
 	}
 
 	// Check for changes
-	if !m.catalogsEqual(m.cached, latestCatalog) {
+	if !m.catalogsEqual(m.hashed, latestCatalog) {
 		// Update cached copy of catalog
-		m.cached = latestCatalog
-
+		m.lock.Lock()
+		m.hashed = latestCatalog
+		m.lock.Unlock()
+		instances, _ := m.registryClient.ListInstances(client.InstanceFilter{})
 		// Notify the listeners.
-		for _, listener := range m.listeners {
-			if err = listener.CatalogChange(latestCatalog); err != nil {
+		m.lock.RLock()
+		copyOfListeners := m.listeners
```

Comment:
Can copy the list of listeners under the w-lock above. It's quicker and simpler than re-acquiring the r-lock.